src/xpcc/architecture/interface/assert.hpp

```diff
@@ -0,0 +1,180 @@
+ * for custom behavior like blinking an LED and printing to a serial connection.
+ *
+ * @warning The abandonment handler may also be executed in interrupt context!
+ *
+ * @see		driver
+ * @author	Niklas Hauser
+ */
+
+namespace xpcc
+{
+
+/// Describes abandonment type of assertions.
+/// @ingroup assert
+enum class
+Abandonment : uint8_t
+{
+	DontCare = Bit0,	///< Do not care about failure.
+	Ignore = Bit1,		///< Safe to ignore this failure.
+	Fail = Bit2			///< This failure is reason for abandonment.
+};
+
+/// Signature of the assertion handlers
+/// @ingroup assert
+using AssertionHandler = Abandonment (*)(const char * module,
+										 const char * location,
+					 					 const char * failure);
+
+} // namespace xpcc
+
+#ifdef __DOXYGEN__
+
+/**
+ * This adds a function to the list of assertion handlers to execute on
+ * assertion failure. Note that this macro does not give you any influence
+ * over the order of handler execution on assertion failure.
+ * Do not write assertion handlers that depend on any ordered execution!
+ *
+ * @warning On AVR targets the failure identifier resides in Flash memory!
+ *
+ * @param handler A function of signature `AssertionHandler`.
+ *
+ * @ingroup assert
+ */
+#define XPCC_ASSERTION_HANDLER(handler)
+
+/**
+ * Assert a condition to be true with failure identifier.
+ * This assert is always included in the source code.
+ *
+ * @note On AVR targets the failure identifier string is placed in Flash memory!
+ *
+ * @ingroup assert
+ */
+#define xpcc_assert(condition, module, location, failure)
+
+/**
+ * Assert a condition to be true with failure identifier.
+ * This assert is only included in the source code on debug builds!
+ *
+ * @note On AVR targets the strings are placed in Flash memory!
+ *
+ * @ingroup assert
+ */
+#define xpcc_assert_debug(condition, module, location, failure)
+
+/**
+ * Overwriteable abandonment handler for all targets.
+ *
+ * You should overwrite this handler for custom failure behaviour like blinking
+ * LEDs and outputting the failure string via a serial connection.
+ *
+ * @ingroup assert
+ */
+extern "C" void
+xpcc_abandon(const char * module,
+			 const char * location,
+			 const char * failure) xpcc_weak;
+
+#else
+
+#if defined XPCC__CPU_ARM || defined XPCC__CPU_AVR
+#	define XPCC_ASSERTION_LINKER_SECTION ".assertion"
+#elif defined XPCC__OS_OSX
+#	define XPCC_ASSERTION_LINKER_SECTION "__DATA,xpcc_assertion"
+#elif defined XPCC__OS_LINUX
+#	define XPCC_ASSERTION_LINKER_SECTION "xpcc_assertion"
+#endif
+
+#ifdef XPCC_ASSERTION_LINKER_SECTION
+#	define XPCC_ASSERTION_HANDLER(handler) \
+		__attribute__((section(XPCC_ASSERTION_LINKER_SECTION), used)) \
+		const xpcc::AssertionHandler \
+		handler ## _assertion_handler_ptr = handler
+#else
+#	define XPCC_ASSERTION_HANDLER(handler)
+#endif
+
+#define xpcc_assert(condition, module, location, failure) \
+	xpcc_assert_evaluate((condition), INLINE_FLASH_STORAGE_STRING(module "\0" location "\0" failure));
```

Comment:
The reason why I chose to put this all into one string is that only one address needs to be loaded into registers. This should make assertions a lot less costly on AVRs, since only 3 registers (2 on ARM) need to be written (one for bool, two for address).